sdk/compiler/daml-lf-tools/src/DA/Daml/LF/TypeChecker/WarnUpgradedDependencies.hs

```diff
@@ -0,0 +1,121 @@
+-- Copyright (c) 2024 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
+-- SPDX-License-Identifier: Apache-2.0
+
+{-# LANGUAGE TypeFamilies #-}
+
+-- | This module provides functions to perform the Daml-LF constraint checks on
+-- types in certain positions. In fact, we also need to do some form of
+-- "constraint inference". To perform this inference in an incremental fashion,
+-- we also provide a way to augment a 'ModuleInterface' with constraint
+-- information about the exported data types in 'augmentInterface'.
+--
+-- Checking whether a function or template definition complies with the Daml-LF
+-- type constraints is straightforward. It is implemented in 'checkModule', which
+-- assumes that the constraint information on the data types in the module being
+-- checked have already been inferred. In other words, the environment must
+-- contain the 'ModuleInterface' produced by 'augmentInterface'.
+module DA.Daml.LF.TypeChecker.WarnUpgradedDependencies
+  ( checkModule
+  ) where
+
+import           Control.Monad.Extra
+import           Data.Foldable (for_)
+
+import DA.Daml.LF.Ast
+import DA.Daml.LF.TypeChecker.Env
+import DA.Daml.LF.TypeChecker.Error
+import Data.Functor.Foldable (cataA, embed)
+import DA.Daml.LF.Ast.Recursive (TypeF(..))
+
+-- | Check whether a type refers to 
+checkType :: forall m. MonadGamma m => Bool -> Type -> m ()
+checkType isSerializable typ = do
+    _ <- cataA go typ
+    pure ()
+  where
+    go :: TypeF (m Type) -> m Type
+    go typeF = do
+      case typeF of
+        TConF tcn -> checkTypeCon isSerializable tcn
+        _ -> pure ()
+      embed <$> sequence typeF
+
+checkTypeCon :: (MonadGamma m) => Bool -> Qualified TypeConName -> m ()
+checkTypeCon isSerializable name = do
+  case qualPackage name of
+    PRSelf -> pure ()
+    PRImport pkgId -> do
+      pkg <- inWorld (lookupPackage (PRImport pkgId))
+
+      -- When using a datatype from a new (>= LF1.17) daml-script in any type
+      -- definition
+      case packageMetadata pkg of
```

Comment:
Move this inside the `isSerializable` check & behind LFVersion check, since we on;y need Daml Script check when using daml script in a serializable location and targeting LF17